<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quantitative features for mass spectrometry data • Features</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Quantitative features for mass spectrometry data">
<meta property="og:description" content="Features">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Features</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Features.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Processing.html">Processing quantitative proteomics data with Features</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RforMassSpectrometry/Features/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quantitative features for mass spectrometry data</h1>
                        <h4 class="author">Laurent Gatto</h4>
                        
      
      <small class="dont-index">Source: <a href="https://github.com/RforMassSpectrometry/Features/blob/master/vignettes/Features.Rmd"><code>vignettes/Features.Rmd</code></a></small>
      <div class="hidden name"><code>Features.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This vignette describes the functionality implemented in the Features package. Features provides infrastructure to manage and process quantitative features for high-throughput mass spectrometry assays, including proteomics and metabolomics.
    </div>
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <code>Features</code> package provides infrastructure (that is classes and the methods to process and manipulate them) to manage and analyse quantitative features from mass spectrometry experiments. It is based on the <code>MultiAssayExperiment</code> class from the <em><a href="https://bioconductor.org/packages/3.11/MultiAssayExperiment">MultiAssayExperiment</a></em> <span class="citation">(Ramos et al. 2017)</span>. that stores a set of assays. Assays in a <code>Features</code> object have a specific relation, that is depicted in figure @ref(fig:featuresplot): assays in a <code>Features</code> object are the result of the aggregation of quantitative features of other assays. In the case of a quantitative proteomics experiment, these different assays would be PSMs, that are aggregated into peptides, that are themselves aggregated into proteins.</p>
<div class="figure">
<img src="Features_files/figure-html/featuresplot-1.png" alt="Conceptual representation of a `Features` object and the aggregative relation between different assays." width="700"><p class="caption">
Conceptual representation of a <code>Features</code> object and the aggregative relation between different assays.
</p>
</div>
<p>In the following sections, we are going to demonstrate how to create a single-assay <code>Features</code> objects starting from a spreadsheet, how to compute the next assays (peptides and proteins), and how these can be manipulated and explored.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"Features"</span>)</pre></body></html></div>
</div>
<div id="creating-features-object" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-features-object" class="anchor"></a>Creating <code>Features</code> object</h1>
<p>While <code>Features</code> objects can be created manually (see <code><a href="../reference/Features-class.html">?Features</a></code> for details), most users will probably possess quantitative data in a spreadsheet or a dataframe. In such cases, the easiest is to use the <code>readFeatures</code> function to extract the quantitative data and metadata columns. Below, we load the <code>hlpsms</code> dataframe that contains data for 28 PSMs from the TMT-10plex <em>hyper</em>LOPIT spatial proteomics experiment from <span class="citation">(Christoforou et al. 2016)</span>. The <code>ecol</code> argument specifies that columns 1 to 10 contain quantitation data, and that the assay should be named <code>psms</code> in the returned <code>Features</code> object, to reflect the nature of the data.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">hlpsms</span>)
<span class="no">hl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/readFeatures.html">readFeatures</a></span>(<span class="no">hlpsms</span>, <span class="kw">ecol</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">10</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">"psms"</span>)
<span class="no">hl</span></pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] psms: SummarizedExperiment with 32776 rows and 10 columns</code></pre>
<p>Below, we see that we can extract an assay using its index or its name. The individual assays are stored as <em>SummerizedExperiment</em> object and further access its quantitative data and metadata using the <code>assay</code> and <code>rowData</code> functions</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">hl</span><span class="kw">[[</span><span class="fl">1</span>]]</pre></body></html></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32776 10 
## metadata(0):
## assays(1): ''
## rownames(32776): 1 2 ... 32775 32776
## rowData names(18): Sequence ProteinDescriptions ... RTmin markers
## colnames(10): X126 X127C ... X130N X131
## colData names(0):</code></pre>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">hl</span><span class="kw">[[</span><span class="st">"psms"</span>]]</pre></body></html></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32776 10 
## metadata(0):
## assays(1): ''
## rownames(32776): 1 2 ... 32775 32776
## rowData names(18): Sequence ProteinDescriptions ... RTmin markers
## colnames(10): X126 X127C ... X130N X131
## colData names(0):</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">assay</span>(<span class="no">hl</span><span class="kw">[[</span><span class="st">"psms"</span>]]))</pre></body></html></div>
<pre><code>##         X126       X127C      X127N       X128C       X128N      X129C
## 1 0.01279427 0.024462641 0.01044012 0.083418628 0.018577277 0.23234391
## 2 0.01524632 0.089010311 0.08151934 0.105754825 0.054639993 0.17449546
## 3 0.13087550 0.023223985 0.01372910 0.010906291 0.010777982 0.10478594
## 4 0.04637540 0.023187698 0.09812058 0.058579448 0.052233341 0.13229192
## 5 0.07011643 0.004831193 0.00321367 0.006790748 0.004389403 0.10047172
## 6 0.05415942 0.127071073 0.05786052 0.366408823 0.113253636 0.04071209
##        X129N      X130C      X130N       X131
## 1 0.29836233 0.02942682 0.16427840 0.12589560
## 2 0.22516965 0.01533445 0.12293998 0.11588966
## 3 0.02951114 0.38663872 0.16038664 0.12916471
## 4 0.15377105 0.09665609 0.17134489 0.16743959
## 5 0.04204136 0.33276803 0.22445809 0.21091935
## 6 0.12275312 0.01020270 0.06057466 0.04700396</code></pre>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">rowData</span>(<span class="no">hl</span><span class="kw">[[</span><span class="st">"psms"</span>]]))</pre></body></html></div>
<pre><code>## DataFrame with 6 rows and 18 columns
##      Sequence ProteinDescriptions NbProteins ProteinGroupAccessions
##   &lt;character&gt;         &lt;character&gt;  &lt;integer&gt;            &lt;character&gt;
## 1     aAAGEAk       Ribosome-b...          1                 Q99PL5
## 2      VTEEAk       Heme oxyge...          1                 P14901
## 3     aSGDPGk       Transcript...          2                 O89090
## 4     sAAAEAk       Isoform 3 ...          3               O89032-3
## 5     aANAEGk       Deoxynucle...          1                 Q8R2M2
## 6     aAAEEAR       Coiled-coi...          1                 Q9JIG7
##   Modifications    qValue       PEP  IonScore NbMissedCleavages
##     &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;integer&gt;
## 1 N-Term(Car...     0.001    0.0162        46                 0
## 2 K6(TMT6ple...     0.002    0.0393        46                 0
## 3 N-Term(Car...     0.008    0.1170        18                 0
## 4 N-Term(Car...     0.001    0.0194        28                 0
## 5 N-Term(Car...     0.002    0.0387        28                 0
## 6 N-Term(TMT...     0.002    0.0386        36                 0
##   IsolationInterference IonInjectTimems Intensity    Charge      mzDa      MHDa
##               &lt;integer&gt;       &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1                     0              70    192000         2   452.258   903.510
## 2                     0              70    416000         2   453.261   905.515
## 3                     0              70    351000         2   459.249   917.491
## 4                     0              70     62100         2   467.263   933.520
## 5                     0              70    246000         2   473.761   946.515
## 6                     0              70    172000         2   473.762   946.516
##   DeltaMassPPM     RTmin       markers
##      &lt;numeric&gt; &lt;numeric&gt;   &lt;character&gt;
## 1        -0.14     10.45 Endoplasmi...
## 2         0.93     15.96       unknown
## 3         1.98     11.42       unknown
## 4        -0.73     13.32       unknown
## 5        -1.05     10.25 Nucleus - ...
## 6         0.88     14.79       unknown</code></pre>
<p>For further details on how to manipulate such objects, refer to the <em><a href="https://bioconductor.org/packages/3.11/MultiAssayExperiment">MultiAssayExperiment</a></em> <span class="citation">(Ramos et al. 2017)</span> and <em><a href="https://bioconductor.org/packages/3.11/SummerizedExperiment">SummerizedExperiment</a></em> <span class="citation">(Morgan et al. 2019)</span> packages.</p>
<p>As illustrated in figure @ref(fig:featuresplot), an central characteristic of <code>Features</code> objects is the aggregative relation between their assays. This can be obtained with the <code>aggregateFeatures</code> function that will aggregate quantitative features from one assay into a new one. In the next code chunk, we aggregate PSM-level data into peptide by grouping all PSMs that were matched the same peptide sequence. Below, the aggregation function is set, as an example, to the mean. The new assay is named <em>peptides</em>.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">hl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Features-aggregate.html">aggregateFeatures</a></span>(<span class="no">hl</span>, <span class="st">"psms"</span>, <span class="st">"Sequence"</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">"peptides"</span>, <span class="kw">fun</span> <span class="kw">=</span> <span class="no">colMeans</span>)
<span class="no">hl</span></pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] psms: SummarizedExperiment with 32776 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 26796 rows and 10 columns</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">hl</span><span class="kw">[[</span><span class="st">"peptides"</span>]]</pre></body></html></div>
<pre><code>## class: SummarizedExperiment 
## dim: 26796 10 
## metadata(0):
## assays(1): ''
## rownames(26796): aAAAAAAAAAAAAAAGAAGk AAAAAAAAAAAAAAAGAAGk ...
##   yYVQPQWVFDcVNAR yYVTIIDAPGHR
## rowData names(7): Sequence ProteinDescriptions ... markers .n
## colnames(10): X126 X127C ... X130N X131
## colData names(0):</code></pre>
<p>Below, we repeat the aggregation operation by grouping peptides into proteins as defined by the <em>ProteinGroupAccessions</em> variable.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">hl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Features-aggregate.html">aggregateFeatures</a></span>(<span class="no">hl</span>, <span class="st">"peptides"</span>, <span class="st">"ProteinGroupAccessions"</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">"proteins"</span>, <span class="kw">fun</span> <span class="kw">=</span> <span class="no">colMeans</span>)
<span class="no">hl</span></pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] psms: SummarizedExperiment with 32776 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 26796 rows and 10 columns 
##  [3] proteins: SummarizedExperiment with 4276 rows and 10 columns</code></pre>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">hl</span><span class="kw">[[</span><span class="st">"proteins"</span>]]</pre></body></html></div>
<pre><code>## class: SummarizedExperiment 
## dim: 4276 10 
## metadata(0):
## assays(1): ''
## rownames(4276): A2A432 A2A4P0 ... Q9Z315 Q9Z321
## rowData names(3): ProteinGroupAccessions markers .n
## colnames(10): X126 X127C ... X130N X131
## colData names(0):</code></pre>
<p>The sample assayed in a <code>Features</code> object can be documented in the <code>colData</code> slot. The <code>hl</code> data doens’t currently possess any sample metadata. These can be addedd as a new <code>DataFrame</code> with matching names (i.e. the <code>DataFrame</code> rownames must be identical assay’s colnames) or can be added one variable at at time, as shown below.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu">colData</span>(<span class="no">hl</span>)</pre></body></html></div>
<pre><code>## DataFrame with 10 rows and 0 columns</code></pre>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">hl</span>$<span class="no">tag</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"126"</span>, <span class="st">"127N"</span>, <span class="st">"127C"</span>, <span class="st">"128N"</span>, <span class="st">"128C"</span>, <span class="st">"129N"</span>, <span class="st">"129C"</span>,
            <span class="st">"130N"</span>, <span class="st">"130C"</span>, <span class="st">"131"</span>)
<span class="fu">colData</span>(<span class="no">hl</span>)</pre></body></html></div>
<pre><code>## DataFrame with 10 rows and 1 column
##               tag
##       &lt;character&gt;
## X126          126
## X127C        127N
## X127N        127C
## X128C        128N
## X128N        128C
## X129C        129N
## X129N        129C
## X130C        130N
## X130N        130C
## X131          131</code></pre>
</div>
<div id="subsetting" class="section level1">
<h1 class="hasAnchor">
<a href="#subsetting" class="anchor"></a>Subsetting</h1>
<p>One particularity of the <code>Features</code> infrastructure is that the features of the constitutive assays are linked through an aggregative relation. This relation is recorded when creating new assays with <code>aggregateFeatures</code> and is exploited when subsetting <code>Feature</code> by their feature names.</p>
<p>In the example below, we are interested in the Stat3B isoform of the Signal transducer and activator of transcription 3 (STAT3) with accession number P42227-2. This accession number corresponds to a feature name in the <em>proteins</em> assay. But this protein row was computed from 8 peptide rows in the <em>peptides</em> assay, themselves resulting from the aggregation of 8 rows in the <em>psms</em> assay.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">stat3</span> <span class="kw">&lt;-</span> <span class="no">hl</span>[<span class="st">"P42227-2"</span>, , ]
<span class="no">stat3</span></pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] proteins: SummarizedExperiment with 1 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 8 rows and 10 columns 
##  [3] psms: SummarizedExperiment with 9 rows and 10 columns</code></pre>
<p>We can easily visualise this new <em>Features</em> object using <code>ggplot2</code> once converted into a <code>data.frame</code>.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">stat3_df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">longFormat</span>(<span class="no">stat3</span>))
<span class="no">stat3_df</span>$<span class="no">assay</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">stat3_df</span>$<span class="no">assay</span>,
                        <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"psms"</span>, <span class="st">"peptides"</span>, <span class="st">"proteins"</span>))

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ggplot2"</span>)
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="no">stat3_df</span>,
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">colname</span>,
           <span class="kw">y</span> <span class="kw">=</span> <span class="no">value</span>,
           <span class="kw">group</span> <span class="kw">=</span> <span class="no">rowname</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(~ <span class="no">assay</span>)</pre></body></html></div>
<p><img src="Features_files/figure-html/plotstat3-1.png" width="700"></p>
<p>Below we repeat the same operation for the Signal transducer and activator of transcription 1 (STAT1) and 3 (STAT3) accession numbers, namely P42227-2 and P42225. We obtain a new <code>Features</code> instance containing 2 proteins, 9 peptides and 10 PSMS. From this, we can readily conclude that STAT1 was identified by a single PSM/peptide.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">stat</span> <span class="kw">&lt;-</span> <span class="no">hl</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"P42227-2"</span>, <span class="st">"P42225"</span>), , ]
<span class="no">stat</span></pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] proteins: SummarizedExperiment with 2 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 9 rows and 10 columns 
##  [3] psms: SummarizedExperiment with 10 rows and 10 columns</code></pre>
<p>Below, we visualise the expression profiles for the two proteins.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">stat_df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">longFormat</span>(<span class="no">stat</span>))
<span class="no">stat_df</span>$<span class="no">stat3</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">stat_df</span>$<span class="no">rowname</span> <span class="kw">%in%</span> <span class="no">stat3_df</span>$<span class="no">rowname</span>,
                        <span class="st">"STAT3"</span>, <span class="st">"STAT1"</span>)
<span class="no">stat_df</span>$<span class="no">assay</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">stat_df</span>$<span class="no">assay</span>,
                        <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"psms"</span>, <span class="st">"peptides"</span>, <span class="st">"proteins"</span>))

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="no">stat_df</span>,
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">colname</span>,
           <span class="kw">y</span> <span class="kw">=</span> <span class="no">value</span>,
           <span class="kw">group</span> <span class="kw">=</span> <span class="no">rowname</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(<span class="no">stat3</span> ~ <span class="no">assay</span>)</pre></body></html></div>
<p><img src="Features_files/figure-html/plotstat-1.png" width="700"></p>
<p>The subsetting by feature names is also available as a call to the <code>subsetByFeature</code> function, for use with the pipe operator.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="no">hl</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="../reference/Features-subsetBy.html">subsetByFeature</a></span>(<span class="st">"P42227-2"</span>)</pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] proteins: SummarizedExperiment with 1 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 8 rows and 10 columns 
##  [3] psms: SummarizedExperiment with 9 rows and 10 columns</code></pre>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">hl</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="../reference/Features-subsetBy.html">subsetByFeature</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"P42227-2"</span>, <span class="st">"P42225"</span>))</pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] proteins: SummarizedExperiment with 2 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 9 rows and 10 columns 
##  [3] psms: SummarizedExperiment with 10 rows and 10 columns</code></pre>
<p>and possibly</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">hl</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="../reference/Features-subsetBy.html">subsetByFeature</a></span>(<span class="st">"P42227-2"</span>) <span class="kw">%&gt;%</span>
    <span class="fu">longFormat</span>() <span class="kw">%&gt;%</span>
    <span class="no">as.data.frame</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">colname</span>,
               <span class="kw">y</span> <span class="kw">=</span> <span class="no">value</span>,
               <span class="kw">group</span> <span class="kw">=</span> <span class="no">rowname</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(~ <span class="no">assay</span>)</pre></body></html></div>
<p>to reproduce the line plot.</p>
<p>Finally, a simply <code>shiny</code> app allows to explore and visualise the respective assays of a <code>Features</code> object.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="fu"><a href="../reference/display.html">display</a></span>(<span class="no">stat</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="figs/display_hmap.png" alt="`Features` interactive interface: heatmap of the peptide assay data." width="100%"><p class="caption">
<code>Features</code> interactive interface: heatmap of the peptide assay data.
</p>
</div>
<div class="figure" style="text-align: center">
<img src="figs/display_assay.png" alt="`Features` interactive interface: quantitative peptide assay data." width="100%"><p class="caption">
<code>Features</code> interactive interface: quantitative peptide assay data.
</p>
</div>
<div class="figure" style="text-align: center">
<img src="figs/display_rowdata.png" alt="`Features` interactive interface: peptide assay row data" width="100%"><p class="caption">
<code>Features</code> interactive interface: peptide assay row data
</p>
</div>
<p>A dropdown menu in the side bar allows the user to select an assay of interest, which can then be visualised as a heatmap (figure @ref(fig:heatmapdisplay)), as a quantitative table (figure @ref(fig:assaydisplay)) or a row data table (figure @ref(fig:rowdatadisplay)).</p>
</div>
<div id="filtering" class="section level1">
<h1 class="hasAnchor">
<a href="#filtering" class="anchor"></a>Filtering</h1>
<p>Features is assays can also be filtered based on variables in their respective row data slots using the <code>filterFeatures</code> function. The filters can be defined using the formula interface or using <code>AnnotationFilter</code> objects from the <em><a href="https://bioconductor.org/packages/3.11/AnnotationFilter">AnnotationFilter</a></em> package <span class="citation">(Morgan and Rainer 2019)</span>. In addition to the pre-defined filters (such as <code>SymbolFilter</code>, <code>ProteinIdFilter</code>, … that filter on gene symbol, protein identifier, …), this package allows users to define arbitrary character or numeric filters using the <code>VariableFilter</code>.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">mito_filter</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Features-filtering.html">VariableFilter</a></span>(<span class="kw">field</span> <span class="kw">=</span> <span class="st">"markers"</span>,
                              <span class="kw">value</span> <span class="kw">=</span> <span class="st">"Mitochondrion"</span>,
                              <span class="kw">condition</span> <span class="kw">=</span> <span class="st">"=="</span>)
<span class="no">mito_filter</span></pre></body></html></div>
<pre><code>## class: CharacterVariableFilter 
## condition: == 
## value: Mitochondrion</code></pre>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">qval_filter</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Features-filtering.html">VariableFilter</a></span>(<span class="kw">field</span> <span class="kw">=</span> <span class="st">"qValue"</span>,
                              <span class="kw">value</span> <span class="kw">=</span> <span class="fl">0.001</span>,
                              <span class="kw">condition</span> <span class="kw">=</span> <span class="st">"&lt;="</span>)
<span class="no">qval_filter</span></pre></body></html></div>
<pre><code>## class: NumericVariableFilter 
## condition: &lt;= 
## value: 0.001</code></pre>
<p>These filter can then readily be applied to all assays’ row data slots. The <code>mito_filter</code> will return all PSMs, peptides and proteins that were annotated as localising to the mitochondrion.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(<span class="no">hl</span>, <span class="no">mito_filter</span>)</pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] psms: SummarizedExperiment with 2019 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 1770 rows and 10 columns 
##  [3] proteins: SummarizedExperiment with 360 rows and 10 columns</code></pre>
<p>The <code>qval_filter</code>, on the other hand, will only return a subset of PSMs, because the <code>qValue</code> variable is only present in the <code>psms</code> assays. The q-values are only relevant to PSMs and that variable was dropped from the other assays.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(<span class="no">hl</span>, <span class="no">qval_filter</span>)</pre></body></html></div>
<pre><code>## harmonizing input:
##   removing 20 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A instance of class Features containing
##  [1] psms: SummarizedExperiment with 26768 rows and 10 columns</code></pre>
<p>The same filters can be created using the forumla interface:</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="fu"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(<span class="no">hl</span>, ~ <span class="no">markers</span> <span class="kw">==</span> <span class="st">"Mitochondrion"</span>)</pre></body></html></div>
<pre><code>## A instance of class Features containing
##  [1] psms: SummarizedExperiment with 2019 rows and 10 columns 
##  [2] peptides: SummarizedExperiment with 1770 rows and 10 columns 
##  [3] proteins: SummarizedExperiment with 360 rows and 10 columns</code></pre>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="fu"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(<span class="no">hl</span>, ~ <span class="no">qValue</span> <span class="kw">&lt;=</span> <span class="fl">0.001</span>)</pre></body></html></div>
<pre><code>## harmonizing input:
##   removing 20 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A instance of class Features containing
##  [1] psms: SummarizedExperiment with 26768 rows and 10 columns</code></pre>
</div>
<div id="session-information" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<pre><code>## R Under development (unstable) (2020-03-13 r77948)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.5 LTS
## 
## Matrix products: default
## BLAS:   /home/travis/R-bin/lib/R/lib/libRblas.so
## LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] magrittr_1.5                 ggplot2_3.3.0               
##  [3] Features_0.4.1               MultiAssayExperiment_1.13.19
##  [5] SummarizedExperiment_1.17.5  DelayedArray_0.13.9         
##  [7] matrixStats_0.56.0           Biobase_2.47.3              
##  [9] GenomicRanges_1.39.3         GenomeInfoDb_1.23.16        
## [11] IRanges_2.21.8               S4Vectors_0.25.15           
## [13] BiocGenerics_0.33.3          BiocStyle_2.15.6            
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.4              lattice_0.20-41         assertthat_0.2.1       
##  [4] rprojroot_1.3-2         digest_0.6.25           R6_2.4.1               
##  [7] backports_1.1.6         evaluate_0.14           highr_0.8              
## [10] pillar_1.4.3            zlibbioc_1.33.1         rlang_0.4.5            
## [13] lazyeval_0.2.2          Matrix_1.2-18           rmarkdown_2.1          
## [16] pkgdown_1.5.0           labeling_0.3            desc_1.2.0             
## [19] stringr_1.4.0           ProtGenerics_1.19.3     RCurl_1.98-1.1         
## [22] munsell_0.5.0           compiler_4.0.0          xfun_0.12              
## [25] pkgconfig_2.0.3         htmltools_0.4.0         tidyselect_1.0.0       
## [28] tibble_3.0.0            GenomeInfoDbData_1.2.2  bookdown_0.18          
## [31] fansi_0.4.1             crayon_1.3.4            dplyr_0.8.5            
## [34] withr_2.1.2             MASS_7.3-51.5           bitops_1.0-6           
## [37] grid_4.0.0              gtable_0.3.0            lifecycle_0.2.0        
## [40] AnnotationFilter_1.11.0 MsCoreUtils_0.99.3      scales_1.1.0           
## [43] cli_2.0.2               stringi_1.4.6           farver_2.0.3           
## [46] XVector_0.27.2          fs_1.3.2                ellipsis_0.3.0         
## [49] vctrs_0.2.4             tools_4.0.0             glue_1.4.0             
## [52] purrr_0.3.3             yaml_2.2.1              colorspace_1.4-1       
## [55] BiocManager_1.30.10     memoise_1.1.0           knitr_1.28.2</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Christoforou:2016">
<p>Christoforou, Andy, Claire M Mulvey, Lisa M Breckels, Aikaterini Geladaki, Tracey Hurrell, Penelope C Hayward, Thomas Naake, et al. 2016. “A Draft Map of the Mouse Pluripotent Stem Cell Spatial Proteome.” <em>Nat Commun</em> 7: 8992. <a href="https://doi.org/10.1038/ncomms9992" class="uri">https://doi.org/10.1038/ncomms9992</a>.</p>
</div>
<div id="ref-SE">
<p>Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2019. <em>SummarizedExperiment: SummarizedExperiment Container</em>.</p>
</div>
<div id="ref-AnnotationFilter">
<p>Morgan, Martin, and Johannes Rainer. 2019. <em>AnnotationFilter: Facilities for Filtering Bioconductor Annotation Resources</em>. <a href="https://github.com/Bioconductor/AnnotationFilter" class="uri">https://github.com/Bioconductor/AnnotationFilter</a>.</p>
</div>
<div id="ref-Ramos:2017">
<p>Ramos, Marcel, Lucas Schiffer, Angela Re, Rimsha Azhar, Azfar Basunia, Carmen Rodriguez Cabrera, Tiffany Chan, et al. 2017. “Software for the Integration of Multi-Omics Experiments in Bioconductor.” <em>Cancer Research</em> 77(21); e39-42.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
