<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Processing quantitative proteomics data with Features • Features</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Processing quantitative proteomics data with Features">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Features</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Features.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Processing.html">Processing quantitative proteomics data with Features</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RforMassSpectrometry/Features">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Processing quantitative proteomics data with Features</h1>
                        <h4 class="author">Laurent Gatto</h4>
                        
      
      <small class="dont-index">Source: <a href="https://github.com/RforMassSpectrometry/Features/blob/master/vignettes/Processing.Rmd"><code>vignettes/Processing.Rmd</code></a></small>
      <div class="hidden name"><code>Processing.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This vignette describes how to process quantitative mass spectrometry data with Features: cleaning up unneeded feature variables, adding an experimental design, filtering out contaminants and reverse hits, managing missing values, log-transforming, normalising and aggregating data.
    </div>
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="reading-data-as-features" class="section level1">
<h1 class="hasAnchor">
<a href="#reading-data-as-features" class="anchor"></a>Reading data as <code>Features</code>
</h1>
<p>We are going to use a subset of the CPTAC study 6 containing conditions A and B <span class="citation">(Paulovich et al. 2010)</span>. The peptide-level data, as processed by MaxQuant <span class="citation">(Cox and Mann 2008)</span> is available in the <code>msdata</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(f &lt;-<span class="st"> </span>msdata<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html">quant</a></span>(<span class="dt">pattern =</span> <span class="st">"cptac"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</a></code></pre></div>
<pre><code>## [1] "cptac_a_b_peptides.txt"</code></pre>
<p>From the names of the columns, we see that the quantitative columns, starting with <code>"Intensity."</code> (note the space!) are at positions 56 to 61.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(f))</a></code></pre></div>
<pre><code>##  [1] "Sequence"                 "N.term.cleavage.window"  
##  [3] "C.term.cleavage.window"   "Amino.acid.before"       
##  [5] "First.amino.acid"         "Second.amino.acid"       
##  [7] "Second.last.amino.acid"   "Last.amino.acid"         
##  [9] "Amino.acid.after"         "A.Count"                 
## [11] "R.Count"                  "N.Count"                 
## [13] "D.Count"                  "C.Count"                 
## [15] "Q.Count"                  "E.Count"                 
## [17] "G.Count"                  "H.Count"                 
## [19] "I.Count"                  "L.Count"                 
## [21] "K.Count"                  "M.Count"                 
## [23] "F.Count"                  "P.Count"                 
## [25] "S.Count"                  "T.Count"                 
## [27] "W.Count"                  "Y.Count"                 
## [29] "V.Count"                  "U.Count"                 
## [31] "Length"                   "Missed.cleavages"        
## [33] "Mass"                     "Proteins"                
## [35] "Leading.razor.protein"    "Start.position"          
## [37] "End.position"             "Unique..Groups."         
## [39] "Unique..Proteins."        "Charges"                 
## [41] "PEP"                      "Score"                   
## [43] "Identification.type.6A_7" "Identification.type.6A_8"
## [45] "Identification.type.6A_9" "Identification.type.6B_7"
## [47] "Identification.type.6B_8" "Identification.type.6B_9"
## [49] "Experiment.6A_7"          "Experiment.6A_8"         
## [51] "Experiment.6A_9"          "Experiment.6B_7"         
## [53] "Experiment.6B_8"          "Experiment.6B_9"         
## [55] "Intensity"                "Intensity.6A_7"          
## [57] "Intensity.6A_8"           "Intensity.6A_9"          
## [59] "Intensity.6B_7"           "Intensity.6B_8"          
## [61] "Intensity.6B_9"           "Reverse"                 
## [63] "Potential.contaminant"    "id"                      
## [65] "Protein.group.IDs"        "Mod..peptide.IDs"        
## [67] "Evidence.IDs"             "MS.MS.IDs"               
## [69] "Best.MS.MS"               "Oxidation..M..site.IDs"  
## [71] "MS.MS.Count"</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">(i &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"Intensity</span><span class="ch">\\</span><span class="st">."</span>, <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(f))))</a></code></pre></div>
<pre><code>## [1] 56 57 58 59 60 61</code></pre>
<p>We now read these data using the <code>readFeatures</code> function. The peptide level expression data will be imported into R as an instance of class <code>Features</code> named <code>cptac</code> with an assay named <code>peptides</code>. We also use the <code>fnames</code> argument to set the row-names of the <code>peptides</code> assay to the peptide sequences.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"Features"</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">cptac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readFeatures.html">readFeatures</a></span>(f, <span class="dt">ecol =</span> i, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">name =</span> <span class="st">"peptides"</span>, <span class="dt">fnames =</span> <span class="st">"Sequence"</span>)</a></code></pre></div>
</div>
<div id="encoding-the-experimental-design" class="section level1">
<h1 class="hasAnchor">
<a href="#encoding-the-experimental-design" class="anchor"></a>Encoding the experimental design</h1>
<p>Below we update the sample (column) annotations to encode the two groups, 6A and 6B, and the original sample numbers.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">cptac<span class="op">$</span>group &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"6A"</span>, <span class="st">"6B"</span>), <span class="dt">each =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">cptac<span class="op">$</span>sample &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">7</span><span class="op">:</span><span class="dv">9</span>, <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="kw">colData</span>(cptac)</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##                      group    sample
##                &lt;character&gt; &lt;integer&gt;
## Intensity.6A_7          6A         7
## Intensity.6A_8          6A         8
## Intensity.6A_9          6A         9
## Intensity.6B_7          6B         7
## Intensity.6B_8          6B         8
## Intensity.6B_9          6B         9</code></pre>
</div>
<div id="filtering-out-contaminants-and-reverse-hits" class="section level1">
<h1 class="hasAnchor">
<a href="#filtering-out-contaminants-and-reverse-hits" class="anchor"></a>Filtering out contaminants and reverse hits</h1>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(cptac, <span class="op">~</span><span class="st"> </span>Reverse <span class="op">==</span><span class="st"> ""</span>)</a></code></pre></div>
<pre><code>## A instance of class Features containing
##  [1] peptides: SummarizedExperiment with 11436 rows and 6 columns</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(cptac, <span class="op">~</span><span class="st"> </span>Potential.contaminant <span class="op">==</span><span class="st"> ""</span>)</a></code></pre></div>
<pre><code>## A instance of class Features containing
##  [1] peptides: SummarizedExperiment with 11385 rows and 6 columns</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"magrittr"</span>)</a>
<a class="sourceLine" id="cb14-2" data-line-number="2">cptac &lt;-<span class="st"> </span>cptac <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="st">    </span><span class="kw"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(<span class="op">~</span><span class="st"> </span>Reverse <span class="op">==</span><span class="st"> ""</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="st">    </span><span class="kw"><a href="../reference/Features-filtering.html">filterFeatures</a></span>(<span class="op">~</span><span class="st"> </span>Potential.contaminant <span class="op">==</span><span class="st"> ""</span>)</a></code></pre></div>
</div>
<div id="removing-up-unneeded-feature-variables" class="section level1">
<h1 class="hasAnchor">
<a href="#removing-up-unneeded-feature-variables" class="anchor"></a>Removing up unneeded feature variables</h1>
<p>The spreadsheet that was read above contained numerous variables that are returned by MaxQuant, but not necessarily necessary in the frame of a downstream statistical analysis.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="../reference/Features-class.html">rowDataNames</a></span>(cptac)</a></code></pre></div>
<pre><code>## CharacterList of length 1
## [["peptides"]] Sequence N.term.cleavage.window ... MS.MS.Count</code></pre>
<p>The only ones that we will be needing below are the peptides sequences and the protein identifiers. Below, we store these variables of interest and filter them using the <code>selectRowData</code> function.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">rowvars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Sequence"</span>, <span class="st">"Proteins"</span>, <span class="st">"Leading.razor.protein"</span>)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">cptac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Features-class.html">selectRowData</a></span>(cptac, rowvars)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="kw"><a href="../reference/Features-class.html">rowDataNames</a></span>(cptac)</a></code></pre></div>
<pre><code>## CharacterList of length 1
## [["peptides"]] Sequence Proteins Leading.razor.protein</code></pre>
</div>
<div id="managing-missing-values" class="section level1">
<h1 class="hasAnchor">
<a href="#managing-missing-values" class="anchor"></a>Managing missing values</h1>
<p>Missing values can be very numerous in certain proteomics experiments and need to be dealt with carefully. The first step is to assess their presence across samples and features. But before being able to do so, we need to replace 0 by <code>NA</code>, given that MaxQuant encodes missing data with a 0 using the <code>zeroIsNA</code> function.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">cptac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Features-missing-data.html">zeroIsNA</a></span>(cptac)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw"><a href="../reference/Features-missing-data.html">nNA</a></span>(cptac)</a></code></pre></div>
<pre><code>## $nNA
## [1] 0.4491943
## 
## $nNArows
## 
##    0    1    2    3    4    5    6 
## 4051  989  879  715  912  797 3014 
## 
## $nNAcols
## Intensity.6A_7 Intensity.6A_8 Intensity.6A_9 Intensity.6B_7 Intensity.6B_8 
##           4669           5388           5224           4651           5470 
## Intensity.6B_9 
##           5207</code></pre>
<p>The output of the <code>nNA</code> function tells us that</p>
<ul>
<li>there are currently close to 50% is missing values in the data;</li>
<li>there are 4051 peptides with 0 missing values, 989 with a single missing values, … and 3014 peptides composed of only missing values;</li>
<li>the range of missing values in the 6 samples is comparable and ranges between 4651 and 5470.</li>
</ul>
<p>In this dataset, we have such a high number of peptides without any data because the 6 samples are a subset of a larger dataset, and these peptides happened to be absent in groups A and B. Below, we use <code>filterNA</code> to remove all the peptides that contain one or more missing values by using <code>pNA = 0</code> (which also is the default value).</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">cptac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Features-missing-data.html">filterNA</a></span>(cptac, <span class="dt">pNA =</span> <span class="dv">0</span>)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">cptac</a></code></pre></div>
<pre><code>## A instance of class Features containing
##  [1] peptides: SummarizedExperiment with 4051 rows and 6 columns</code></pre>
<p>I we wanted to keep peptides that have up to 90% of missing values, corresponsing in this case to those that have only one value (i.e 5/6 percent of missing values), we could have set <code>pNA</code> to 0.9.</p>
</div>
<div id="imputation" class="section level1">
<h1 class="hasAnchor">
<a href="#imputation" class="anchor"></a>Imputation</h1>
<p>The <code>impute</code> method can be used to perform missing value imputation using a variety of imputation methods. The method takes an instance of class <code>Features</code> (or a <code>SummarizedExperiment</code>) as input, an a character naming the desired method (see <code><a href="../reference/impute.html">?impute</a></code> for the complete list with details) and returns a new instance of class <code>Features</code> (or <code>SummarizedExperiment</code>) with imputed data.</p>
<p>As described in more details in <span class="citation">(Lazar et al. 2016)</span>, there are two types of mechanisms resulting in missing values in LC/MSMS experiments.</p>
<ul>
<li><p>Missing values resulting from absence of detection of a feature, despite ions being present at detectable concentrations. For example in the case of ion suppression or as a result from the stochastic, data-dependent nature of the MS acquisition method. These missing value are expected to be randomly distributed in the data and are defined as <em>missing at random</em> (MAR) or <em>missing completely at random</em> (MCAR).</p></li>
<li><p>Biologically relevant missing values, resulting from the <em>absence</em> of the low abundance of ions (below the limit of detection of the instrument). These missing values are not expected to be randomly distributed in the data and are defined as <em>missing not at random</em> (MNAR).</p></li>
</ul>
<p>MAR and MCAR values can be reasonably well tackled by many imputation methods. MNAR data, however, requires some knowledge about the underlying mechanism that generates the missing data, to be able to attempt data imputation. MNAR features should ideally be imputed with a <em>left-censor</em> (for example using a deterministic or probabilistic minimum value) method. Conversely, it is recommended to use <em>hot deck</em> methods (for example nearest neighbour, maximum likelihood, etc) when data are missing at random.</p>
<div class="figure">
<img src="Processing_files/figure-html/miximp-1.png" alt="Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation." width="700"><p class="caption">
Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation.
</p>
</div>
<p>It is anticipated that the identification of both classes of missing values will depend on various factors, such as feature intensities and experimental design. Below, we use perform mixed imputation, applying nearest neighbour imputation on the 654 features that are assumed to contain randomly distributed missing values (if any) (yellow on figure @ref(fig:miximp)) and a deterministic minimum value imputation on the 35 proteins that display a non-random pattern of missing values (brown on figure @ref(fig:miximp)).</p>
</div>
<div id="data-transformation" class="section level1">
<h1 class="hasAnchor">
<a href="#data-transformation" class="anchor"></a>Data transformation</h1>
<p>When analysing continuous data using parametric methods (such as t-test or linear models), it is often necessary to log-transform the data. The figure below (left) show that how our data is mainly composed of small values with a long tail of larger ones, which is a typical pattern of quantitative omics data.</p>
<p>Below, we use the <code>logTransform</code> function to log2-transform our data. This time, instead of overwriting the peptides assay, we are going to create a new one to contain the log2-transformed data.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">cptac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Features-class.html">addAssay</a></span>(cptac,</a>
<a class="sourceLine" id="cb23-2" data-line-number="2">                  <span class="kw"><a href="../reference/Features-processing.html">logTransform</a></span>(cptac[[<span class="dv">1</span>]]),</a>
<a class="sourceLine" id="cb23-3" data-line-number="3">                  <span class="dt">name =</span> <span class="st">"peptides_log"</span>)</a>
<a class="sourceLine" id="cb23-4" data-line-number="4">cptac</a></code></pre></div>
<pre><code>## A instance of class Features containing
##  [1] peptides: SummarizedExperiment with 4051 rows and 6 columns 
##  [2] peptides_log: SummarizedExperiment with 4051 rows and 6 columns</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb25-2" data-line-number="2">limma<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span>(<span class="kw">assay</span>(cptac[[<span class="dv">1</span>]]))</a>
<a class="sourceLine" id="cb25-3" data-line-number="3">limma<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span>(<span class="kw">assay</span>(cptac[[<span class="dv">2</span>]]))</a></code></pre></div>
<div class="figure">
<img src="Processing_files/figure-html/unnamed-chunk-10-1.png" alt="Quantitative data in its original scale (left) and log2-transformed (right)." width="700"><p class="caption">
Quantitative data in its original scale (left) and log2-transformed (right).
</p>
</div>
</div>
<div id="normalisation" class="section level1">
<h1 class="hasAnchor">
<a href="#normalisation" class="anchor"></a>Normalisation</h1>
<p>Assays in <code>Features</code> objects can be normalised with the <code>normalize</code> function. The type of normalisation is defined by the <code>method</code> argument; below, we use quantiles normalisation, store the normalised data into a new experiment, and visualise the resulting data.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">cptac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Features-class.html">addAssay</a></span>(cptac,</a>
<a class="sourceLine" id="cb26-2" data-line-number="2">                  <span class="kw"><a href="../reference/Features-processing.html">normalize</a></span>(cptac[[<span class="st">"peptides_log"</span>]], <span class="dt">method =</span> <span class="st">"quantiles"</span>),</a>
<a class="sourceLine" id="cb26-3" data-line-number="3">                  <span class="dt">name =</span> <span class="st">"peptides_norm"</span>)</a></code></pre></div>
<pre><code>## Loading required namespace: preprocessCore</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">cptac</a></code></pre></div>
<pre><code>## A instance of class Features containing
##  [1] peptides: SummarizedExperiment with 4051 rows and 6 columns 
##  [2] peptides_log: SummarizedExperiment with 4051 rows and 6 columns 
##  [3] peptides_norm: SummarizedExperiment with 4051 rows and 6 columns</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb30-2" data-line-number="2">limma<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span>(<span class="kw">assay</span>(cptac[[<span class="st">"peptides_log"</span>]]))</a>
<a class="sourceLine" id="cb30-3" data-line-number="3">limma<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/limma/man/plotDensities.html">plotDensities</a></span>(<span class="kw">assay</span>(cptac[[<span class="st">"peptides_norm"</span>]]))</a></code></pre></div>
<div class="figure">
<img src="Processing_files/figure-html/unnamed-chunk-12-1.png" alt="Distribution of log2 peptide intensities before (left) and after (right) quantiles normalisation." width="700"><p class="caption">
Distribution of log2 peptide intensities before (left) and after (right) quantiles normalisation.
</p>
</div>
</div>
<div id="feature-aggregation" class="section level1">
<h1 class="hasAnchor">
<a href="#feature-aggregation" class="anchor"></a>Feature aggregation</h1>
<p>At this stage, it is possible to directly use the peptide-level intensities to perform a statistical analysis <span class="citation">(Goeminne, Gevaert, and Clement 2016)</span>, or aggregate the peptide-level data into protein intensities, and perform the differential expression analysis at the protein level.</p>
<p>To aggregate feature data, we can use the <code>aggregateFeatures</code> function that takes the following inputs:</p>
<ul>
<li>the name of the <code>Features</code> instance that contains the peptide quantitation data - <code>"cptac"</code> in our example;</li>
<li>
<strong><code>i</code></strong>: the name or index of the assay that contains the (normalised) peptide quantitation data - <code>"peptides_norm"</code> in our case;</li>
<li>
<strong><code>fcol</code></strong>: the feature variable (in the assay above) to be used to define what peptides to aggregate - <code>"Proteins"</code> here, given that we want to aggregate all peptides that belong to one protein (group);</li>
<li>
<strong><code>name</code></strong>: the name of the new aggregates assay - <code>"proteins"</code> in this case;</li>
<li>and finally <strong><code>fun</code></strong>, the function that will compute this aggregation - we will be using the default value, namely <code>robustSummary</code> <span class="citation">(Sticker et al. 2019)</span>.</li>
</ul>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1">cptac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Features-aggregate.html">aggregateFeatures</a></span>(cptac, <span class="dt">i =</span> <span class="st">"peptides_norm"</span>, <span class="dt">fcol =</span> <span class="st">"Proteins"</span>, <span class="dt">name =</span> <span class="st">"proteins"</span>)</a>
<a class="sourceLine" id="cb31-2" data-line-number="2">cptac</a></code></pre></div>
<pre><code>## A instance of class Features containing
##  [1] peptides: SummarizedExperiment with 4051 rows and 6 columns 
##  [2] peptides_log: SummarizedExperiment with 4051 rows and 6 columns 
##  [3] peptides_norm: SummarizedExperiment with 4051 rows and 6 columns 
##  [4] proteins: SummarizedExperiment with 1125 rows and 6 columns</code></pre>
<p>We obtain a final 1125 quantified proteins in the new <code>proteins</code> assay. Below, we display the quantitation data for the first 6 proteins and their respective variables. The latter shown that number of peptides that were using during the aggregation step (<code>.n</code> column).</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">assay</span>(cptac[[<span class="st">"proteins"</span>]]))</a></code></pre></div>
<pre><code>##                                      Intensity.6A_7 Intensity.6A_8
## P00918ups|CAH2_HUMAN_UPS                   17.30380       16.89828
## P01008ups|ANT3_HUMAN_UPS;CON__P41361       16.89434       15.98417
## P01127ups|PDGFB_HUMAN_UPS                  16.52022       16.81015
## P02144ups|MYG_HUMAN_UPS                    16.88925       16.46800
## P02753ups|RETBP_HUMAN_UPS                  17.84540       16.65529
## P02787ups|TRFE_HUMAN_UPS                   16.82634       16.88629
##                                      Intensity.6A_9 Intensity.6B_7
## P00918ups|CAH2_HUMAN_UPS                   16.63440       18.28533
## P01008ups|ANT3_HUMAN_UPS;CON__P41361       16.30053       16.82910
## P01127ups|PDGFB_HUMAN_UPS                  16.83638       18.20670
## P02144ups|MYG_HUMAN_UPS                    17.27137       17.87518
## P02753ups|RETBP_HUMAN_UPS                  16.49888       18.39589
## P02787ups|TRFE_HUMAN_UPS                   16.34263       18.14808
##                                      Intensity.6B_8 Intensity.6B_9
## P00918ups|CAH2_HUMAN_UPS                   18.57310       18.47074
## P01008ups|ANT3_HUMAN_UPS;CON__P41361       16.72590       16.44177
## P01127ups|PDGFB_HUMAN_UPS                  18.79476       17.14791
## P02144ups|MYG_HUMAN_UPS                    18.58260       18.29267
## P02753ups|RETBP_HUMAN_UPS                  17.73285       18.13718
## P02787ups|TRFE_HUMAN_UPS                   18.52184       18.15342</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="kw">rowData</span>(cptac[[<span class="st">"proteins"</span>]])</a></code></pre></div>
<pre><code>## DataFrame with 1125 rows and 3 columns
##                                           Proteins Leading.razor.protein
##                                        &lt;character&gt;           &lt;character&gt;
## P00918ups|CAH2_HUMAN_UPS             P00918ups|...         P00918ups|...
## P01008ups|ANT3_HUMAN_UPS;CON__P41361 P01008ups|...         P01008ups|...
## P01127ups|PDGFB_HUMAN_UPS            P01127ups|...         P01127ups|...
## P02144ups|MYG_HUMAN_UPS              P02144ups|...         P02144ups|...
## P02753ups|RETBP_HUMAN_UPS            P02753ups|...         P02753ups|...
## ...                                            ...                   ...
## sp|Q99207|NOP14_YEAST                sp|Q99207|...         sp|Q99207|...
## sp|Q99216|PNO1_YEAST                 sp|Q99216|...         sp|Q99216|...
## sp|Q99257|MEX67_YEAST                sp|Q99257|...         sp|Q99257|...
## sp|Q99258|RIB3_YEAST                 sp|Q99258|...         sp|Q99258|...
## sp|Q99383|HRP1_YEAST                 sp|Q99383|...         sp|Q99383|...
##                                             .n
##                                      &lt;integer&gt;
## P00918ups|CAH2_HUMAN_UPS                     1
## P01008ups|ANT3_HUMAN_UPS;CON__P41361         1
## P01127ups|PDGFB_HUMAN_UPS                    1
## P02144ups|MYG_HUMAN_UPS                      1
## P02753ups|RETBP_HUMAN_UPS                    2
## ...                                        ...
## sp|Q99207|NOP14_YEAST                        1
## sp|Q99216|PNO1_YEAST                         1
## sp|Q99257|MEX67_YEAST                        2
## sp|Q99258|RIB3_YEAST                         2
## sp|Q99383|HRP1_YEAST                         2</code></pre>
<p>We can get a quick overview of this <code>.n</code> variable by computing the table below, that shows us that we have 405 proteins that are based on a single peptides, 230 that are based on two, 119 that are based on three, … and a single protein that is the results of aggregating 44 peptides.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="kw">rowData</span>(cptac[[<span class="st">"proteins"</span>]])<span class="op">$</span>.n)</a></code></pre></div>
<pre><code>## 
##   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 
## 405 230 119  84  64  53  37  29  24  24  13   9   4   3   3   7   3   1   1   1 
##  21  22  23  24  25  30  31  33  44 
##   1   2   2   1   1   1   1   1   1</code></pre>
<p>Let’s choose <code>P02787ups|TRFE_HUMAN_UPS</code> and visualise its expression pattern in the 2 groups at the protein and peptide level.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ggplot2"</span>)</a>
<a class="sourceLine" id="cb39-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"dplyr"</span>)</a>
<a class="sourceLine" id="cb39-3" data-line-number="3"><span class="kw">longFormat</span>(cptac[<span class="st">"P02787ups|TRFE_HUMAN_UPS"</span>, ]) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb39-4" data-line-number="4"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb39-5" data-line-number="5"><span class="st">    </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">group =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"A"</span>, colname), <span class="st">"A"</span>, <span class="st">"B"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb39-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">sample =</span> <span class="kw"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"Intensity</span><span class="ch">\\</span><span class="st">."</span>, <span class="st">""</span>, colname)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb39-7" data-line-number="7"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> sample, <span class="dt">y =</span> value, <span class="dt">colour =</span> rowname, <span class="dt">shape =</span> group)) <span class="op">+</span></a>
<a class="sourceLine" id="cb39-8" data-line-number="8"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb39-9" data-line-number="9"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(<span class="op">~</span><span class="st"> </span>assay)</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 12 sampleMap rows not in names(experiments)
## harmonizing input:
##   removing 12 sampleMap rows not in names(experiments)</code></pre>
<div class="figure">
<img src="Processing_files/figure-html/unnamed-chunk-16-1.png" alt="Expression intensities for the protein *P02787ups|TRFE_HUMAN_UPS* (right, green) and its peptides (left) in groups A (circles) and B (triangles)." width="700"><p class="caption">
Expression intensities for the protein <em>P02787ups|TRFE_HUMAN_UPS</em> (right, green) and its peptides (left) in groups A (circles) and B (triangles).
</p>
</div>
</div>
<div id="todo" class="section level1">
<h1 class="hasAnchor">
<a href="#todo" class="anchor"></a>TODO</h1>
<ul>
<li>Improve on data visualisation.</li>
</ul>
</div>
<div id="session-information" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<pre><code>## R Under development (unstable) (2020-02-06 r77774)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.5 LTS
## 
## Matrix products: default
## BLAS:   /home/travis/R-bin/lib/R/lib/libRblas.so
## LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] gplots_3.0.1.2              magrittr_1.5               
##  [3] dplyr_0.8.4                 ggplot2_3.2.1              
##  [5] Features_0.4.0              MultiAssayExperiment_1.13.6
##  [7] SummarizedExperiment_1.17.1 DelayedArray_0.13.3        
##  [9] BiocParallel_1.21.2         matrixStats_0.55.0         
## [11] Biobase_2.47.2              GenomicRanges_1.39.2       
## [13] GenomeInfoDb_1.23.10        IRanges_2.21.3             
## [15] S4Vectors_0.25.12           BiocGenerics_0.33.0        
## [17] BiocStyle_2.15.6           
## 
## loaded via a namespace (and not attached):
##  [1] gtools_3.8.1            assertthat_0.2.1        BiocManager_1.30.10    
##  [4] highr_0.8               msdata_0.27.0           GenomeInfoDbData_1.2.2 
##  [7] yaml_2.2.1              pillar_1.4.3            backports_1.1.5        
## [10] lattice_0.20-38         glue_1.3.1              limma_3.43.3           
## [13] digest_0.6.23           XVector_0.27.0          colorspace_1.4-1       
## [16] htmltools_0.4.0         preprocessCore_1.49.2   Matrix_1.2-18          
## [19] plyr_1.8.5              pkgconfig_2.0.3         bookdown_0.17          
## [22] zlibbioc_1.33.1         purrr_0.3.3             scales_1.1.0           
## [25] gdata_2.18.0            tibble_2.1.3            farver_2.0.3           
## [28] AnnotationFilter_1.11.0 withr_2.1.2             lazyeval_0.2.2         
## [31] crayon_1.3.4            memoise_1.1.0           evaluate_0.14          
## [34] fs_1.3.1                MASS_7.3-51.5           tools_4.0.0            
## [37] lifecycle_0.1.0         stringr_1.4.0           munsell_0.5.0          
## [40] compiler_4.0.0          pkgdown_1.4.1           caTools_1.18.0         
## [43] rlang_0.4.4             grid_4.0.0              RCurl_1.98-1.1         
## [46] MsCoreUtils_0.99.0      bitops_1.0-6            labeling_0.3           
## [49] rmarkdown_2.1           gtable_0.3.0            reshape2_1.4.3         
## [52] R6_2.4.1                knitr_1.28.1            rprojroot_1.3-2        
## [55] ProtGenerics_1.19.3     KernSmooth_2.23-16      desc_1.2.0             
## [58] stringi_1.4.5           Rcpp_1.0.3              tidyselect_1.0.0       
## [61] xfun_0.12</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Cox:2008">
<p>Cox, J, and M Mann. 2008. “MaxQuant Enables High Peptide Identification Rates, Individualized P.p.b.-range Mass Accuracies and Proteome-Wide Protein Quantification.” <em>Nat Biotechnol</em> 26 (12): 1367–72. <a href="https://doi.org/10.1038/nbt.1511" class="uri">https://doi.org/10.1038/nbt.1511</a>.</p>
</div>
<div id="ref-Goeminne:2016">
<p>Goeminne, L J, K Gevaert, and L Clement. 2016. “Peptide-Level Robust Ridge Regression Improves Estimation, Sensitivity, and Specificity in Data-Dependent Quantitative Label-Free Shotgun Proteomics.” <em>Mol Cell Proteomics</em> 15 (2): 657–68. <a href="https://doi.org/10.1074/mcp.M115.055897" class="uri">https://doi.org/10.1074/mcp.M115.055897</a>.</p>
</div>
<div id="ref-Lazar:2016">
<p>Lazar, C, L Gatto, M Ferro, C Bruley, and T Burger. 2016. “Accounting for the Multiple Natures of Missing Values in Label-Free Quantitative Proteomics Data Sets to Compare Imputation Strategies.” <em>J Proteome Res</em> 15 (4): 1116–25. <a href="https://doi.org/10.1021/acs.jproteome.5b00981" class="uri">https://doi.org/10.1021/acs.jproteome.5b00981</a>.</p>
</div>
<div id="ref-Paulovich:2010">
<p>Paulovich, Amanda G, Dean Billheimer, Amy-Joan L Ham, Lorenzo Vega-Montoto, Paul A Rudnick, David L Tabb, Pei Wang, et al. 2010. “Interlaboratory Study Characterizing a Yeast Performance Standard for Benchmarking LC-MS Platform Performance.” <em>Mol. Cell. Proteomics</em> 9 (2): 242–54.</p>
</div>
<div id="ref-Sticker:2019">
<p>Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement. 2019. “Robust Summarization and Inference in Proteome-Wide Label-Free Quantification.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/668863" class="uri">https://doi.org/10.1101/668863</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#reading-data-as-features">Reading data as <code>Features</code></a></li>
      <li><a href="#encoding-the-experimental-design">Encoding the experimental design</a></li>
      <li><a href="#filtering-out-contaminants-and-reverse-hits">Filtering out contaminants and reverse hits</a></li>
      <li><a href="#removing-up-unneeded-feature-variables">Removing up unneeded feature variables</a></li>
      <li><a href="#managing-missing-values">Managing missing values</a></li>
      <li><a href="#imputation">Imputation</a></li>
      <li><a href="#data-transformation">Data transformation</a></li>
      <li><a href="#normalisation">Normalisation</a></li>
      <li><a href="#feature-aggregation">Feature aggregation</a></li>
      <li><a href="#todo">TODO</a></li>
      <li><a href="#session-information">Session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
